### PR TITLE
Add custom derives for ULE and VarULE

### PR DESCRIPTION
**Pull request by Manishearth**
This is a step towards https://github.com/unicode-org/icu4x/issues/1079, implementing [the first part of the design doc](https://github.com/unicode-org/icu4x/blob/main/utils/zerovec/design_doc.md#deriveule-and-derivevarule)

Note that this is _not_ the final form of this feature: `#[derive(ULE)]` and `#[derive(VarULE)]` are building blocks that most users will not directly use. Rather, these custom derives are for the specialized use case where you wish to implement your own AsULE/EncodeAsVarULE conversions.

I have not started using this in the codebase, but this could be already be used to simplify the RelationULE code if so desired (I'd rather just make that change when the final attribute is ready).

I've also not set up the reexports from the zerovec crate, we can do that at the end (and perhaps reorganize the ULE/VarULE types to be reexported?)

A thought I'm having is that it's worth writing fuzzers for these things because it's easy to write tests that ensure that parsing correct data works, but harder to ensure that incorrect data is rejected.

cc @robertbastian, @echeran, @nordzilla in case you are interested in looking at this to follow along.

### CODE REVIEW

File: utils/zerovec/derive/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# This file is part of ICU4X. For terms of use, please see the file
+# called LICENSE at the top level of the ICU4X source tree
+# (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+[package]
+name = "zerovec-derive"
```
Comment (sffc) [inline utils/zerovec/derive/Cargo.toml:6]: Question: `zerovec-derive` or `zerovec_derive` ?

File: utils/zerovec/derive/src/ule.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        || input.generics.const_params().next().is_some()
+    {
+        return Error::new(
+            input.generics.span(),
+            "derive(ULE) must be applied to a struct without any generics",
+        )
+        .to_compile_error();
+    }
+    let struc = if let Data::Struct(ref s) = input.data {
+        if s.fields.iter().next().is_none() {
```
Comment (sffc) [inline utils/zerovec/derive/src/ule.rs:35]: Thought: I noticed that one of the `yoke` derives (not sure which) fails on this condition (it doesn't work on an empty struct).  Can we add this condition there, too?

File: utils/zerovec/derive/src/ule.rs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    };
+
+    let (validators, remaining_offset) = generate_ule_validators(struc.fields.iter());
+
+    let name = &input.ident;
+
+    // Safety (based on the safety checklist on the ULE trait):
+    //  1. #name does not include any uninitialized or padding bytes.
+    //     (achieved by enforcing #[repr(transparent)] or #[repr(packed)] on a struct of only ULE types)
+    //  2. CharULE is aligned to 1 byte.
```
Comment (sffc) [inline utils/zerovec/derive/src/ule.rs:55]: Nit: `CharULE`

File: utils/zerovec/derive/src/ule.rs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    //     (achieved by enforcing #[repr(transparent)] or #[repr(packed)] on a struct of only ULE types)
+    //  3. The impl of validate_byte_slice() returns an error if any byte is not valid.
+    //  4. The impl of validate_byte_slice() returns an error if there are extra bytes.
+    //  5. The other ULE methods use the default impl.
+    //  6. [This impl does not enforce the equality constraint, it is up to the user to do so, ideally via a custom derive]
```
Comment (sffc) [inline utils/zerovec/derive/src/ule.rs:60]: Thought: Do we get the equality constraint by way of wrapping `ULE` types that also satisfy the constraint?

File: utils/zerovec/derive/src/varule.rs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        )
+        .to_compile_error();
+    }
+    if input.generics.type_params().next().is_some()
+        || input.generics.lifetimes().next().is_some()
+        || input.generics.const_params().next().is_some()
+    {
+        return Error::new(
+            input.generics.span(),
+            "derive(ULE) must be applied to a struct without any generics",
```
Comment (sffc) [inline utils/zerovec/derive/src/varule.rs:25]: Nit, here and elsewhere: `VarULE`

**Review by sffc — APPROVED** [review]
Cool!

File: utils/zerovec/derive/src/ule.rs (hunk 4 above)
Comment (Manishearth) [reply utils/zerovec/derive/src/ule.rs:60]: only if people derive partialeq

File: utils/zerovec/derive/Cargo.toml (hunk 1 above)
Comment (Manishearth) [reply utils/zerovec/derive/Cargo.toml:6]: I typically use dashes and the yoke derive crate does the same

File: utils/zerovec/derive/src/ule.rs (hunk 2 above)
Comment (Manishearth) [reply utils/zerovec/derive/src/ule.rs:35]: I think both of them should be fine with empty fields? Note that you can't even have a lifetime with zero fields so it always goes through the simpler lts=0 path.

**Review by sffc — APPROVED** [review]

**Pull request merged**